python/ray/_private/utils.py

```
@@ -801,6 +802,97 @@ def set_sigterm_handler(sigterm_handler):
         signal.signal(signal.SIGTERM, sigterm_handler)
 
 
+_signal_handler_installed = False
+_graceful_shutdown_in_progress = False
+
+
+def install_driver_signal_handler() -> None:
+    """Install SIGTERM handler for Ray driver processes.
+
+    Implements graceful-then-forced shutdown semantics:
+    - First SIGTERM: trigger graceful shutdown via sys.exit() (allows atexit handlers)
+    - Second SIGTERM: escalate to immediate forced shutdown via os._exit(1)
+
+    Must be called from the main thread (Python signal handlers requirement).
+    Refer to https://docs.python.org/3/library/signal.html#signals-and-threads for more details.
+    """
+    global _signal_handler_installed, _graceful_shutdown_in_progress
+    if _signal_handler_installed:
+        return
+
+    if threading.current_thread() is not threading.main_thread():
+        logger.warning(
+            "Signal handlers not installed because current thread is not the main thread. Refer to https://docs.python.org/3/library/signal.html#signals-and-threads for more details."
+        )
+        return
```

This should raise an exception. Does it happen anywhere today?